### PR TITLE
Add Contract Flattener

### PR DESCRIPTION
**Pull request by mateuszradomski**
Resolves L2B-3730

- [x] Unit tests
- [x] Integration tests
- [x] Try to compile some of the flattened files and compare the bytecode
	- Socket -> Socket
	- Base -> OptimismPortal
	- Base -> L2OutputOracle
	- dydx -> CpuFrilessVerifier [here](https://etherscan.io/address/0x4922f8750DFd040954b44F23980160342e308863#code)
- [x] Refactor the way sources are saved in the discovery to allow for easier integration with the flattener
- [x] Tests and build

### CODE REVIEW

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+      const areTopLevelPresent =
+        file.rootASTNode.children.filter((n) => n.type === 'FunctionDefinition')
+          .length !== 0
+      assert(!areTopLevelPresent, 'Function definitions are not supported')
+    }
+
+    return result
+  }
+
+  resolveContractDeclarations(file: ParsedFile): ContractDeclaration[] {
```
Comment (sz-piotr) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:124]: private

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    )
+    assert(matchingContract !== undefined, 'Contract not found')
+
+    return {
+      contract: matchingContract,
+      file,
+    }
+  }
+
+  resolveImportPath(fromFile: ParsedFile, importPath: string): ParsedFile {
```
Comment (sz-piotr) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:307]: private

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        referencedContracts: [],
+        byteRange: {
+          start: c.range[0],
+          end: c.range[1],
+        },
+      }
+    })
+  }
+
+  resolveFileImports(
```
Comment (sz-piotr) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:149]: private

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+        }
+      }
+
+      alreadyImportedObjects.set(importedFile.path, alreadyImported)
+
+      return result
+    })
+  }
+
+  resolveReferencedLibraries(
```
Comment (sz-piotr) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:227]: private

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+  inheritsFrom: string[]
+  referencedContracts: string[]
+}
+
+// If import is:
+//
+// import { foo as bar } from 'baz'
+//
+// Then:
+// path: 'baz'
```
Comment (sz-piotr) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:34]: this comment is out of date. path != absolutePath.

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+    )
+
+    const result = posix.join(longest.target, path.slice(longest.prefix.length))
+    return result
+  }
+
+  return path
+}
+
+function findOne<T>(
```
Comment (sz-piotr) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:360]: why not use `Array.prototype.find`?

**Review by sz-piotr — CHANGES_REQUESTED** [review]

**Review by adamiak — COMMENTED** [review]
First of all I want to say good job! The way this source code is written is uncomparabaly more readable and easy to follow than the original prototype. Finding proper place to add new functionality should not be a problem for anyone. Same goes for fixing bugs.

I've left some comments in the code, but here I want to share some of my high level thoughts regarding structure of the solution. I hope this will be useful in the future.

1. First of all, the parsing solution is mostly written as a single God-like object: `ParsedFilesManager`. This approach misses many chances of breaking the problem into smaller, self-contained pieces of logic.

For example, `ParsedFilesManager` has these methods: 

- `resolveContractDeclarations(file)`, 
- `tryFindContract(contractName, file)`, 
- `resolveReferencedLibraries(file, contract)`.

These methods operate on a single "file". Why is the logic regarding single file stuffed into ParsedFilesManager and not into a separate class like. e.g `File` or `ParsedFile`? All above functions could be methods of such class and wouldn't need "file" passed as parameter.

Whenever you see yourself passing the same argument to many methods of a class, it might be a sign that they should be part of a separate class (and the argument becomes `.this`).

2. The point above can be applied further. Parts of logic that operate on a Contract should be inside some "Contract" class, while right now they are spread between many methods in `ParsedFilesManager` and `flattenStartingFrom.ts` file. 

3. Because of putting so much code into ParsedFilesManager, nothing prevented you from spreading the code relating to e.g. a single file throught the whole class.

For example, in `ParsedFilesManager.parseFiles()` you first create a list of all files, but with mostly empty bodies ({ `contractDeclarations: [], importDirectives: []` }. And then you use multiple "passes" of going throught all files and filling those empty bodies, first with contract declarations, then with imports and so on. It's a large function with multiple loops which inside operate mostly on a single file. 

If you had a separate "File" class (`ParsedFile`?), then instead of four separate loops of building all files iteratively , you would have a single loop, asking each file to build/instantiate itself.

As before, if you write a code that operates mostly on a single object (e.g. `.file`), it might be better to separate this logic out. 

4. Since everything is kept in one big `ParsedFileManager`, the flattening logic that's extracted  to `flattenStartingFrom.ts` feels like another method, artificially pulled outside. It accesses all the data in `ParsedFilesManager` and calls `tryFindContract()` over and over anyway. You also create these unique contract SHA hashes which seem like something that should rather live in a "ParsedContract" class if it existed.

If you instead followed the approach I outlined above and broke the code into separate classes, like:

- `ParsedFilesManager`
- `ParsedFile`
- `ParsedContract`
- potentially other, like `ParsedLibrary` etc.

it would follow naturally to put `.flatten` method into each of them, or maybe to just `ParsedContract` (which would internally flatted its base contracts), nicely separating it into well defined pieces. And even if you preferred to put logic into separate Flattener class, its implementation would be much clearer - the whole "SHA unique ids" should not be necessary at all since e.g. each contract would have a single instance of `ParsedContract` class.

To summarize, readability is very good and flow of the code is easy to follow. I think the logic could be divided into smaller pieces, which would make both flattening and parsing logic even clearer. Overall, really good job and an amazing tool!

File: packages/discovery/src/flatten/getASTIdentifiers.ts [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+  TypeName,
+  UncheckedStatement,
+  UsingForDeclaration,
+  VariableDeclaration,
+  VariableDeclarationStatement,
+  WhileStatement,
+} from '@solidity-parser/parser/dist/src/ast-types'
+/* eslint-enable */
+
+export function getASTIdentifiers(node: BaseASTNode | null): string[] {
```
Comment (adamiak) [inline packages/discovery/src/flatten/getASTIdentifiers.ts:37]: I have a feeling that his whole file is an overkill for what we want to achieve. I don't think we need *all* the identifiers - we only need 3 or 4 (contract, library, function, import?) and I guess we can gather them from wherever they are in the AST, without need to walk the whole tree and understanding every single structure.

Also, see the comment added to the test file.

File: packages/discovery/src/flatten/getASTIdentifiers.test.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        'UsedLibrary1',
+        'Contract1',
+        'EventLibrary.EventType',
+        'emitIndex',
+        'ErrorLibrary.ErrorType',
+        'errorCode',
+        'Lib.Buffer',
+        'buffers',
+        'ParamLib1.Param',
+        'param1',
```
Comment (adamiak) [inline packages/discovery/src/flatten/getASTIdentifiers.test.ts:60]: I think that pulling identifiers this way can potentially be harmful via name clashes, or mostly not useful. What do we gain by having "param1" as identifier? Or "bytes"? We really care only about a few identifier types (contracts, libs, functions, imports). From this list we don't even know what each identifier is, meaning we need to have additional code somewhere else to prevent name clashes. And if we don't have it, then we can easily misinterpret a function call from param.

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+      return result
+    })
+  }
+
+  resolveReferencedLibraries(
+    file: ParsedFile,
+    c: ContractDefinition,
+  ): string[] {
+    const identifiers = new Set(
+      c.subNodes.flatMap((n) => getASTIdentifiers(n)).map(extractNamespace),
```
Comment (adamiak) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:232]: This is the only place we use `getASTIdentifiers()`. Do we really need all the `getASTIdentifiers()` complexity to pull referenced libraries?

File: packages/discovery/src/flatten/ParsedFilesManager.ts [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    )
+
+    const referenced = []
+    for (const identifier of identifiers) {
+      const result = this.tryFindContract(identifier, file)
```
Comment (adamiak) [inline packages/discovery/src/flatten/ParsedFilesManager.ts:237]: What about name clashes? If there's a parameter to some function that has the same name as some contracts, won't there be a false positive?

**Review by adamiak — APPROVED** [review]
As discussed, I'm approving to get this out for researchers to play with and provide feedback. Please look at the comments and introduce improvements where necessary via new, small PRs (rather than a big refactoring).

File: packages/discovery/src/flatten/ParsedFilesManager.ts (hunk 6 above)
Comment (mateuszradomski) [reply packages/discovery/src/flatten/ParsedFilesManager.ts:363]: Because we want to fail if we find more than one.

File: packages/discovery/src/discovery/output/saveDiscoveryResult.ts [hunk 11]
```diff
@@ -63,6 +102,141 @@ export async function saveDiscoveryResult(
...
+): string {
+  assert(contract.type !== 'EOA', 'Invalid execution path')
+
+  const hasProxy = contract.source.length > 1
+  const isProxy = hasProxy && sourceIndex === 0
+
+  const hasNameClash =
+    allContractNames.filter((n) => n === source.name).length > 1
+  const uniquenessSuffix =
+    hasNameClash && !hasProxy ? `-${source.address.toString()}` : ''
```
[the remark this answers is not in the document]
Comment (mateuszradomski) [reply packages/discovery/src/discovery/output/saveDiscoveryResult.ts:205]: It will never happen since proxies have a suffix `.p.sol`. So if there is a proxy named exactly the same as a contract the suffix will save it from clashing

File: packages/discovery/src/discovery/output/saveDiscoveryResult.ts [hunk 12]
```diff
@@ -63,6 +102,141 @@ export async function saveDiscoveryResult(
...
+function getFlatSourceFileName(
+  contract: Analysis,
+  sourceIndex: number,
+  source: PerContractSource,
+  allContractNames: string[],
+): string {
+  assert(contract.type !== 'EOA', 'Invalid execution path')
+
+  const hasProxy = contract.source.length > 1
+  const isProxy = hasProxy && sourceIndex === 0
```
[the remark this answers is not in the document]
Comment (mateuszradomski) [reply packages/discovery/src/discovery/output/saveDiscoveryResult.ts:200]: The scheme we came up with is somewhat complex, so the implementing the logic in a clean way was hard. This is the most readable I was able to make it. I'm open to improvement to logic/scheme if you think it's too complex.

File: packages/discovery/src/discovery/output/saveDiscoveryResult.ts [hunk 13]
```diff
@@ -63,6 +102,141 @@ export async function saveDiscoveryResult(
...
+
+  logger.log(`Saving flattened sources`)
+  for (const contract of results) {
+    try {
+      if (contract.type === 'EOA') {
+        continue
+      }
+
+      for (const [sourceIndex, source] of contract.source.entries()) {
+        const input: FileContent[] = Object.entries(source.files)
```
[the remark this answers is not in the document]
Comment (mateuszradomski) [reply packages/discovery/src/discovery/output/saveDiscoveryResult.ts:131]: Good catch, it's already fixed. It will skip this source, logging that it happened and informing the user that it's because we found zero `.sol` files.

File: packages/discovery/src/discovery/output/saveDiscoveryResult.ts [hunk 14]
```diff
@@ -63,6 +102,141 @@ export async function saveDiscoveryResult(
...
+  logger: DiscoveryLogger,
+  options: {
+    rootFolder?: string
+    sourcesFolder?: string
+    discoveryFilename?: string
+  },
+): Promise<void> {
+  const flatSourcesFolder = `.flat${options.sourcesFolder ?? ''}`
+  const flatSourcesPath = posix.join(rootPath, flatSourcesFolder)
+  const allContractNames = results.map((c) =>
```
[the remark this answers is not in the document]
Comment (mateuszradomski) [reply packages/discovery/src/discovery/output/saveDiscoveryResult.ts:117]: I agree that saving sources grew too much. I wanted to refactor it but decided against it to cut development time. I'll plan and talk about separating it into three steps instead of two.

Linear issue ID: [L2B-4185](https://linear.app/l2beat/issue/L2B-4185/simplify-the-architecture-of-saving-source-files-in-discovery)
